metadata/src/from/v16.rs

```diff
@@ -0,0 +1,260 @@
+                let name: ArcStr = c.name.clone().into();
+                (name.clone(), from_constant_metadata(name, c))
+            });
+            let view_functions = p
+                .view_functions
+                .into_iter()
+                .map(from_view_function_metadata);
+
+            let call_variant_index = VariantIndex::build(p.calls.as_ref().map(|c| c.ty.id), &types);
+            let error_variant_index =
+                VariantIndex::build(p.error.as_ref().map(|e| e.ty.id), &types);
+            let event_variant_index =
+                VariantIndex::build(p.event.as_ref().map(|e| e.ty.id), &types);
+
+            let associated_types = p
+                .associated_types
+                .into_iter()
+                .map(|t| (t.name, t.ty.id))
+                .collect();
+
+            pallets_by_index.insert(p.index, pos);
+            pallets.push_insert(
+                name.clone(),
+                PalletMetadataInner {
+                    name,
+                    index: p.index,
+                    storage,
+                    call_ty: p.calls.map(|c| c.ty.id),
+                    call_variant_index,
+                    event_ty: p.event.map(|e| e.ty.id),
+                    event_variant_index,
+                    error_ty: p.error.map(|e| e.ty.id),
+                    error_variant_index,
+                    constants: constants.collect(),
+                    view_functions: view_functions.collect(),
+                    associated_types,
+                    docs: p.docs,
+                },
+            );
+        }
+
+        let apis = m.apis.into_iter().map(|api| {
+            let name: ArcStr = api.name.clone().into();
+            (name.clone(), from_runtime_api_metadata(name, api))
+        });
+
+        let custom_map = m
+            .custom
+            .map
+            .into_iter()
+            .map(|(key, val)| {
+                let custom_val = v15::CustomValueMetadata {
+                    ty: val.ty,
+                    value: val.value,
+                };
+                (key, custom_val)
+            })
+            .collect();
+
+        let dispatch_error_ty = types
+            .types
+            .iter()
+            .find(|ty| ty.ty.path.segments == ["sp_runtime", "DispatchError"])
+            .map(|ty| ty.id);
+
+        Ok(Metadata {
+            types,
+            pallets,
+            pallets_by_index,
+            extrinsic: from_extrinsic_metadata(m.extrinsic),
+            dispatch_error_ty,
+            apis: apis.collect(),
+            outer_enums: OuterEnumsMetadata {
+                call_enum_ty: m.outer_enums.call_enum_ty.id,
+                event_enum_ty: m.outer_enums.event_enum_ty.id,
+                error_enum_ty: m.outer_enums.error_enum_ty.id,
+            },
+            custom: v15::CustomMetadata { map: custom_map },
+        })
+    }
+}
+
+fn from_transaction_extension_metadata(
```

Comment:
same dq about `From` instances

Reply:
I'd write these as `From` impls if I wanted it a part of a public interface, but I suppose I didn't want that sort of interface so it was easier to just write freestanding functions to do what was needed :)